std/sumtype.d

```diff
@@ -2575,6 +2575,35 @@ version (D_Exceptions)
     }
 }
 
+// return ref
+@safe unittest
+{
+    static assert(!__traits(compiles, () {
+
+        SumType!(int, string) st;
+        return st.match!(
+            (string x) => imported!"std.typecons".NullableRef!int.init,
+            (return ref int i) => imported!"std.typecons".nullableRef!int(&i),
```

Comment:
Is there a reason we're using `NullableRef` here? The example in the bug report returns a plain `ref`.

Reply:
I wanted to show that it correctly follows the lifetime even if sent to another instance. But this is most likely testing `dip1000` instead of the `SumType` itself.

Reply:
More to the point, it's testing `NullableRef`, which means this test can fail if there's bug in *either* `SumType` or `NullableRef`.

Reply:
Yeah, I can remove this dependency and leave only the plain `ref` or create a small `struct` that takes a pointer.

Reply:
Thanks. Plain `ref` like in the bug report would be preferable.

Reply:
Oops, I just realized the example in the bug report is wrong, it's supposed to be the same as the last unittest:

```d
SumType!(int, string) st;
st.match!(
  function int*(string _) => assert(0),
  function int*(ref int i) => &i,
);
```

It doesn't make sense for `ref` to be there :confounded: 
I will fix the bug report example, but the explanation remains the same.

Reply:
Done, and added a unittest for a type with indirections. I also left a comment explaining that although that code wouldn't work in this version, the same was already accomplished by the removal of `ref` as it does nothing in this context. I don't know if it brings much value or not, what do you think?

Reply:
I think the new test case is more confusing than helpful (it took me a long time to figure it out—the correct fix is to remove `return`, not `ref`), and it doesn't test anything that isn't already tested by the previous examples. My vote is to remove it.

Reply:
done